File: server/src/auth/auth.service.ts
import {Body, HttpException, HttpStatus, Injectable, Post, UnauthorizedException} from '@nestjs/common';
import {CreateUserDto} from "../users/dto/create-user.dto";
import {UsersService} from "../users/users.service";
import {JwtService} from "@nestjs/jwt";

import * as bcrypt from 'bcrypt';
import {User} from "../users/schemas/user.schema";

@Injectable()
export class AuthService {

    constructor(private userService: UsersService,
                private jwtService: JwtService) {}

    async login(userDto: CreateUserDto) {
        const user = await this.validateUser(userDto)
        return this.generateToken(user)
    }

    async registration(userDto: CreateUserDto) {
        const candidate = await this.userService.getUserByEmail(userDto.email);
        const candidate2 = await this.userService.getUserByUsername(userDto.username);
        if (candidate) {
            throw new HttpException('user with this email already exists', HttpStatus.BAD_REQUEST);
        } else if(candidate2){
            throw new HttpException('user with this name already exists', HttpStatus.BAD_REQUEST);
        }
        const hashPassword = await bcrypt.hash(userDto.password, 5);
        const user = await this.userService.createUser({...userDto, password: hashPassword})
        return this.generateToken(user)
    }

    private async generateToken(user: User) {
        const userId = await this.userService.getUserByEmail(user.email);
        const payload = {id: userId._id, name: user.username, email: user.email};
        return {
            token: this.jwtService.sign(payload)
        }
    }

    private async validateUser(userDto: CreateUserDto) {
        const user = await this.userService.getUserByEmail(userDto.email);
        const passwordEquals = await bcrypt.compare(userDto.password, user.password);
        if (user && passwordEquals) {
            return user;
        }
        throw new UnauthorizedException({message: 'incorrect email or password'})
    }
}

File: server-java/src/main/java/com/muz/service/AuthService.java
package com.muz.service;

import com.muz.document.UserDocument;
import com.muz.dto.CreateUserDto;
import com.muz.dto.LoginUserDto;
import com.muz.dto.TokenResponseDto;
import com.muz.exception.ApiException;
import com.muz.security.JwtUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserService userService;
    private final JwtUtil jwtUtil;
    private final PasswordEncoder passwordEncoder;

    public TokenResponseDto login(LoginUserDto dto) {
        UserDocument user = userService.getByEmail(dto.getEmail());
        if (user == null || !passwordEncoder.matches(dto.getPassword(), user.getPassword())) {
            throw new ApiException(HttpStatus.UNAUTHORIZED, "incorrect email or password");
        }
        return new TokenResponseDto(jwtUtil.generateToken(user.getId(), user.getUsername(), user.getEmail()));
    }

    public TokenResponseDto registration(CreateUserDto dto) {
        if (userService.getByEmail(dto.getEmail()) != null) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "user with this email already exists");
        }
        if (userService.getByUsername(dto.getUsername()) != null) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "user with this name already exists");
        }
        dto.setPassword(passwordEncoder.encode(dto.getPassword()));
        UserDocument user = userService.createUser(dto);
        return new TokenResponseDto(jwtUtil.generateToken(user.getId(), user.getUsername(), user.getEmail()));
    }
}
